File: OpenHarmony/developtools/hapsigner/hapsigntool_cpp/zip/src/central_directory.cpp
/*
 * Copyright (c) 2024-2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "central_directory.h"
#include "signature_tools_log.h"

namespace OHOS {
namespace SignatureTools {
bool CentralDirectory::GetCentralDirectory(ByteBuffer& bf, CentralDirectory* cd)
{
    int signatureValue;
    bf.GetInt32(signatureValue);
    if (signatureValue != SIGNATURE) {
        SIGNATURE_TOOLS_LOGE("find zip central directory failed");
        return false;
    }

    SetCentralDirectoryValues(bf, cd);

    uint16_t fileNameLength = cd->GetFileNameLength();
    if (fileNameLength > 0) {
        std::string readFileName(fileNameLength, 0);
        bf.GetData(&readFileName[0], fileNameLength);
        cd->SetFileName(readFileName);
    }
    uint16_t extraLength = cd->GetExtraLength();
    if (extraLength > 0) {
        std::string extra(extraLength, 0);
        bf.GetData(&extra[0], extraLength);
        cd->SetExtraData(extra);
    }
    uint16_t commentLength = cd->GetCommentLength();
    if (commentLength > 0) {
        std::string readComment(commentLength, 0);
        bf.GetData(&readComment[0], commentLength);
        cd->SetComment(readComment);
    }
    cd->SetLength(CD_LENGTH + fileNameLength + extraLength + commentLength);

    return true;
}

void CentralDirectory::SetCentralDirectoryValues(ByteBuffer& bf, CentralDirectory* cd)
{
    int16_t centralDirectoryInt16Value;
    bf.GetInt16(centralDirectoryInt16Value);
    cd->SetVersion(centralDirectoryInt16Value);

    bf.GetInt16(centralDirectoryInt16Value);
    cd->SetVersionExtra(centralDirectoryInt16Value);

    bf.GetInt16(centralDirectoryInt16Value);
    cd->SetFlag(centralDirectoryInt16Value);

    bf.GetInt16(centralDirectoryInt16Value);
    cd->SetMethod(centralDirectoryInt16Value);

    bf.GetInt16(centralDirectoryInt16Value);
    cd->SetLastTime(centralDirectoryInt16Value);

    bf.GetInt16(centralDirectoryInt16Value);
    cd->SetLastDate(centralDirectoryInt16Value);

    int32_t centralDirectoryInt32Value;
    bf.GetInt32(centralDirectoryInt32Value);
    cd->SetCrc32(centralDirectoryInt32Value);

    uint32_t centralDirectoryUInt32Value;
    bf.GetUInt32(centralDirectoryUInt32Value);
    cd->SetCompressedSize(centralDirectoryUInt32Value);

    bf.GetUInt32(centralDirectoryUInt32Value);
    cd->SetUnCompressedSize(centralDirectoryUInt32Value);

    uint16_t centralDirectoryUInt16Value;
    bf.GetUInt16(centralDirectoryUInt16Value);
    cd->SetFileNameLength(centralDirectoryUInt16Value);

    bf.GetUInt16(centralDirectoryUInt16Value);
    cd->SetExtraLength(centralDirectoryUInt16Value);

    bf.GetUInt16(centralDirectoryUInt16Value);
    cd->SetCommentLength(centralDirectoryUInt16Value);

    bf.GetUInt16(centralDirectoryUInt16Value);
    cd->SetDiskNumStart(centralDirectoryUInt16Value);

    bf.GetInt16(centralDirectoryInt16Value);
    cd->SetInternalFile(centralDirectoryInt16Value);

    bf.GetInt32(centralDirectoryInt32Value);
    cd->SetExternalFile(centralDirectoryInt32Value);

    bf.GetUInt32(centralDirectoryUInt32Value);
    cd->SetOffset(centralDirectoryUInt32Value);
}

std::string CentralDirectory::ToBytes()
{
    ByteBuffer bf(m_length);
    bf.PutInt32(SIGNATURE);
    bf.PutInt16(m_version);
    bf.PutInt16(m_versionExtra);
    bf.PutInt16(m_flag);
    bf.PutInt16(m_method);
    bf.PutInt16(m_lastTime);
    bf.PutInt16(m_lastDate);
    bf.PutInt32(m_crc32);
    bf.PutUInt32(m_compressedSize);
    bf.PutUInt32(m_unCompressedSize);
    bf.PutUInt16(m_fileNameLength);
    bf.PutUInt16(m_extraLength);
    bf.PutUInt16(m_commentLength);
    bf.PutUInt16(m_diskNumStart);
    bf.PutInt16(m_internalFile);
    bf.PutInt32(m_externalFile);
    bf.PutUInt32(m_offset);

    if (m_fileNameLength > 0) {
        bf.PutData(m_fileName.c_str(), m_fileName.size());
    }
    if (m_extraLength > 0) {
        bf.PutData(m_extraData.c_str(), m_extraData.size());
    }
    if (m_commentLength > 0) {
        bf.PutData(m_extraData.c_str(), m_extraData.size());
    }

    return bf.ToString();
}

int CentralDirectory::GetCdLength()
{
    return CD_LENGTH;
}

int CentralDirectory::GetSIGNATURE()
{
    return SIGNATURE;
}

short CentralDirectory::GetVersion()
{
    return m_version;
}

void CentralDirectory::SetVersion(short version)
{
    m_version = version;
}

short CentralDirectory::GetVersionExtra()
{
    return m_versionExtra;
}

void CentralDirectory::SetVersionExtra(short versionExtra)
{
    m_versionExtra = versionExtra;
}

short CentralDirectory::GetFlag()
{
    return m_flag;
}

void CentralDirectory::SetFlag(short flag)
{
    m_flag = flag;
}

short CentralDirectory::GetMethod()
{
    return m_method;
}

void CentralDirectory::SetMethod(short method)
{
    m_method = method;
}

short CentralDirectory::GetLastTime()
{
    return m_lastTime;
}

void CentralDirectory::SetLastTime(short lastTime)
{
    m_lastTime = lastTime;
}

short CentralDirectory::GetLastDate()
{
    return m_lastDate;
}

void CentralDirectory::SetLastDate(short lastDate)
{
    m_lastDate = lastDate;
}

int CentralDirectory::GetCrc32()
{
    return m_crc32;
}

void CentralDirectory::SetCrc32(int crc32)
{
    m_crc32 = crc32;
}

uint32_t CentralDirectory::GetCompressedSize()
{
    return m_compressedSize;
}

void CentralDirectory::SetCompressedSize(uint32_t compressedSize)
{
    m_compressedSize = compressedSize;
}

uint32_t CentralDirectory::GetUnCompressedSize()
{
    return m_unCompressedSize;
}

void CentralDirectory::SetUnCompressedSize(uint32_t unCompressedSize)
{
    m_unCompressedSize = unCompressedSize;
}

uint16_t CentralDirectory::GetFileNameLength()
{
    return m_fileNameLength;
}

void CentralDirectory::SetFileNameLength(uint16_t fileNameLength)
{
    m_fileNameLength = fileNameLength;
}

uint16_t CentralDirectory::GetExtraLength()
{
    return m_extraLength;
}

void CentralDirectory::SetExtraLength(uint16_t extraLength)
{
    m_extraLength = extraLength;
}

uint16_t CentralDirectory::GetCommentLength()
{
    return m_commentLength;
}

void CentralDirectory::SetCommentLength(uint16_t commentLength)
{
    m_commentLength = commentLength;
}

uint16_t CentralDirectory::GetDiskNumStart()
{
    return m_diskNumStart;
}

void CentralDirectory::SetDiskNumStart(uint16_t diskNumStart)
{
    m_diskNumStart = diskNumStart;
}

short CentralDirectory::GetInternalFile()
{
    return m_internalFile;
}

void CentralDirectory::SetInternalFile(short internalFile)
{
    m_internalFile = internalFile;
}

int CentralDirectory::GetExternalFile()
{
    return m_externalFile;
}

void CentralDirectory::SetExternalFile(int externalFile)
{
    m_externalFile = externalFile;
}

uint32_t CentralDirectory::GetOffset()
{
    return m_offset;
}

void CentralDirectory::SetOffset(uint32_t offset)
{
    m_offset = offset;
}

std::string CentralDirectory::GetFileName()
{
    return m_fileName;
}

void CentralDirectory::SetFileName(const std::string& fileName)
{
    m_fileName = fileName;
}

std::string CentralDirectory::GetExtraData() const
{
    return m_extraData;
}

void CentralDirectory::SetExtraData(const std::string& extraData)
{
    m_extraData = extraData;
}

std::string CentralDirectory::GetComment()
{
    return m_comment;
}

void CentralDirectory::SetComment(const std::string& comment)
{
    m_comment = comment;
}

uint32_t CentralDirectory::GetLength()
{
    return m_length;
}

void CentralDirectory::SetLength(uint32_t length)
{
    m_length = length;
}
} // namespace SignatureTools
} // namespace OHOS

File: OpenHarmony/developtools/hapsigner/hapsigntool/hap_sign_tool_lib/src/main/java/com/ohos/hapsigntool/zip/CentralDirectory.java
/*
 * Copyright (c) 2023-2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool.zip;

import com.ohos.hapsigntool.error.ZipException;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;

/**
 * resolve zip CentralDirectory data
 * CentralDirectory format for:
 * central file header signature   4 bytes  (0x02014b50)
 * version made by                 2 bytes
 * version needed to extract       2 bytes
 * general purpose bit flag        2 bytes
 * compression method              2 bytes
 * last mod file time              2 bytes
 * last mod file date              2 bytes
 * crc-32                          4 bytes
 * compressed size                 4 bytes
 * uncompressed size               4 bytes
 * file name length                2 bytes
 * extra field length              2 bytes
 * file comment length             2 bytes
 * disk number start               2 bytes
 * internal file attributes        2 bytes
 * external file attributes        4 bytes
 * relative offset of local header 4 bytes
 * file name (variable size)
 * extra field (variable size)
 * file comment (variable size)
 *
 * @since 2023/12/02
 */
public class CentralDirectory {
    /**
     * central directory invariable bytes length
     */
    public static final int CD_LENGTH = 46;

    /**
     * 4 bytes , central directory signature
     */
    public static final int SIGNATURE = 0x02014b50;

    /**
     * 2 bytes
     */
    private short version;

    /**
     * 2 bytes
     */
    private short versionExtra;

    /**
     * 2 bytes
     */
    private short flag;

    /**
     * 2 bytes
     */
    private short method;

    /**
     * 2 bytes
     */
    private short lastTime;

    /**
     * 2 bytes
     */
    private short lastDate;

    /**
     * 4 bytes
     */
    private int crc32;

    /**
     * 4 bytes
     */
    private long compressedSize;

    /**
     * 4 bytes
     */
    private long unCompressedSize;

    /**
     * 2 bytes
     */
    private int fileNameLength;

    /**
     * 2 bytes
     */
    private int extraLength;

    /**
     * 2 bytes
     */
    private int commentLength;

    /**
     * 2 bytes
     */
    private int diskNumStart;

    /**
     * 2 bytes
     */
    private short internalFile;

    /**
     * 4 bytes
     */
    private int externalFile;

    /**
     * 4 bytes
     */
    private long offset;

    /**
     * n bytes
     */
    private String fileName;

    /**
     * n bytes
     */
    private byte[] extraData;

    /**
     * n bytes
     */
    private byte[] comment;

    private int length;

    /**
     * updateLength
     */
    public void updateLength() {
        length = CD_LENGTH + fileNameLength + extraLength + commentLength;
    }

    /**
     * get Central Directory
     *
     * @param bf ByteBuffer
     * @return CentralDirectory
     * @throws ZipException read Central Directory exception
     */
    public static CentralDirectory getCentralDirectory(ByteBuffer bf) throws ZipException {
        CentralDirectory cd = new CentralDirectory();
        if (bf.getInt() != SIGNATURE) {
            throw new ZipException("find zip central directory failed");
        }

        cd.setVersion(bf.getShort());
        cd.setVersionExtra(bf.getShort());
        cd.setFlag(bf.getShort());
        cd.setMethod(bf.getShort());
        cd.setLastTime(bf.getShort());
        cd.setLastDate(bf.getShort());
        cd.setCrc32(bf.getInt());
        cd.setCompressedSize(UnsignedDecimalUtil.getUnsignedInt(bf));
        cd.setUnCompressedSize(UnsignedDecimalUtil.getUnsignedInt(bf));
        cd.setFileNameLength(UnsignedDecimalUtil.getUnsignedShort(bf));
        cd.setExtraLength(UnsignedDecimalUtil.getUnsignedShort(bf));
        cd.setCommentLength(UnsignedDecimalUtil.getUnsignedShort(bf));
        cd.setDiskNumStart(UnsignedDecimalUtil.getUnsignedShort(bf));
        cd.setInternalFile(bf.getShort());
        cd.setExternalFile(bf.getInt());
        cd.setOffset(UnsignedDecimalUtil.getUnsignedInt(bf));
        if (cd.getFileNameLength() > 0) {
            byte[] readFileName = new byte[cd.getFileNameLength()];
            bf.get(readFileName);
            cd.setFileName(new String(readFileName, StandardCharsets.UTF_8));
        }
        if (cd.getExtraLength() > 0) {
            byte[] extra = new byte[cd.getExtraLength()];
            bf.get(extra);
            cd.setExtraData(extra);
        }
        if (cd.getCommentLength() > 0) {
            byte[] readComment = new byte[cd.getCommentLength()];
            bf.get(readComment);
            cd.setComment(readComment);
        }
        cd.updateLength();
        return cd;
    }

    /**
     * change Central Directory to bytes
     *
     * @return bytes
     */
    public byte[] toBytes() {
        ByteBuffer bf = ByteBuffer.allocate(length).order(ByteOrder.LITTLE_ENDIAN);
        bf.putInt(SIGNATURE);
        UnsignedDecimalUtil.setUnsignedShort(bf, version);
        UnsignedDecimalUtil.setUnsignedShort(bf, versionExtra);
        UnsignedDecimalUtil.setUnsignedShort(bf, flag);
        UnsignedDecimalUtil.setUnsignedShort(bf, method);
        UnsignedDecimalUtil.setUnsignedShort(bf, lastTime);
        UnsignedDecimalUtil.setUnsignedShort(bf, lastDate);
        UnsignedDecimalUtil.setUnsignedInt(bf, crc32);
        UnsignedDecimalUtil.setUnsignedInt(bf, compressedSize);
        UnsignedDecimalUtil.setUnsignedInt(bf, unCompressedSize);
        UnsignedDecimalUtil.setUnsignedShort(bf, fileNameLength);
        UnsignedDecimalUtil.setUnsignedShort(bf, extraLength);
        UnsignedDecimalUtil.setUnsignedShort(bf, commentLength);
        UnsignedDecimalUtil.setUnsignedShort(bf, diskNumStart);
        UnsignedDecimalUtil.setUnsignedShort(bf, internalFile);
        UnsignedDecimalUtil.setUnsignedInt(bf, externalFile);
        UnsignedDecimalUtil.setUnsignedInt(bf, offset);
        if (fileNameLength > 0) {
            bf.put(fileName.getBytes(StandardCharsets.UTF_8));
        }
        if (extraLength > 0) {
            bf.put(extraData);
        }
        if (commentLength > 0) {
            bf.put(extraData);
        }
        return bf.array();
    }

    public static int getCdLength() {
        return CD_LENGTH;
    }

    public static int getSIGNATURE() {
        return SIGNATURE;
    }

    public short getVersion() {
        return version;
    }

    public void setVersion(short version) {
        this.version = version;
    }

    public short getVersionExtra() {
        return versionExtra;
    }

    public void setVersionExtra(short versionExtra) {
        this.versionExtra = versionExtra;
    }

    public short getFlag() {
        return flag;
    }

    public void setFlag(short flag) {
        this.flag = flag;
    }

    public short getMethod() {
        return method;
    }

    public void setMethod(short method) {
        this.method = method;
    }

    public short getLastTime() {
        return lastTime;
    }

    public void setLastTime(short lastTime) {
        this.lastTime = lastTime;
    }

    public short getLastDate() {
        return lastDate;
    }

    public void setLastDate(short lastDate) {
        this.lastDate = lastDate;
    }

    public int getCrc32() {
        return crc32;
    }

    public void setCrc32(int crc32) {
        this.crc32 = crc32;
    }

    public long getCompressedSize() {
        return compressedSize;
    }

    public void setCompressedSize(long compressedSize) {
        this.compressedSize = compressedSize;
    }

    public long getUnCompressedSize() {
        return unCompressedSize;
    }

    public void setUnCompressedSize(long unCompressedSize) {
        this.unCompressedSize = unCompressedSize;
    }

    public int getFileNameLength() {
        return fileNameLength;
    }

    public void setFileNameLength(int fileNameLength) {
        this.fileNameLength = fileNameLength;
    }

    public int getExtraLength() {
        return extraLength;
    }

    public void setExtraLength(int extraLength) {
        this.extraLength = extraLength;
    }

    public int getCommentLength() {
        return commentLength;
    }

    public void setCommentLength(int commentLength) {
        this.commentLength = commentLength;
    }

    public int getDiskNumStart() {
        return diskNumStart;
    }

    public void setDiskNumStart(int diskNumStart) {
        this.diskNumStart = diskNumStart;
    }

    public short getInternalFile() {
        return internalFile;
    }

    public void setInternalFile(short internalFile) {
        this.internalFile = internalFile;
    }

    public int getExternalFile() {
        return externalFile;
    }

    public void setExternalFile(int externalFile) {
        this.externalFile = externalFile;
    }

    public long getOffset() {
        return offset;
    }

    public void setOffset(long offset) {
        this.offset = offset;
    }

    public String getFileName() {
        return fileName;
    }

    public void setFileName(String fileName) {
        this.fileName = fileName;
    }

    public byte[] getExtraData() {
        return extraData;
    }

    public void setExtraData(byte[] extraData) {
        this.extraData = extraData;
    }

    public byte[] getComment() {
        return comment;
    }

    public void setComment(byte[] comment) {
        this.comment = comment;
    }

    public int getLength() {
        return length;
    }

    public void setLength(int length) {
        this.length = length;
    }
}